### PR TITLE
[AMDGPU][True16][MC] VOP3 profile in True16 format

### PR DESCRIPTION
**Pull request by broxigarchen**
Modify VOP3 profile and pesudo, and add encoding info for VOP3 True16 including DPP and DPP8 in true16 and fake16 format.

This patch applies true16/fake16 changes and asm/dasm changes to
V_ADD_NC_U16
V_ADD_NC_I16
V_SUB_NC_U16
V_SUB_NC_I16

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td [hunk 1]
```diff
@@ -760,9 +756,14 @@ class OpSelBinOpClampPat<SDPatternOperator node,
   (inst $src0_modifiers, $src0, $src1_modifiers, $src1, DSTCLAMP.ENABLE, 0)
 >;
 
-def : OpSelBinOpClampPat<saddsat, V_ADD_I16_e64>;
-def : OpSelBinOpClampPat<ssubsat, V_SUB_I16_e64>;
-} // End SubtargetPredicate = isGFX9Plus
+let OtherPredicates = [isGFX9Plus], True16Predicate = NotHasTrue16BitInsts in {
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/VOP3Instructions.td:759]:
```suggestion
let SubtargetPredicate = isGFX9Plus, True16Predicate = NotHasTrue16BitInsts in {
```

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td [hunk 2]
```diff
@@ -871,21 +872,31 @@ let SubtargetPredicate = isGFX10Plus in {
...
-  def : GCNPat<
-    (add i16:$src0, (i16 NegSubInlineIntConst16:$src1)),
-    (V_SUB_NC_U16_e64 0, VSrc_b16:$src0, 0, NegSubInlineIntConst16:$src1, 0, 0)
-  >;
+  defm V_ADD_NC_U16 : VOP3Inst_t16 <"v_add_nc_u16", VOP_I16_I16_I16, add>;
+  defm V_SUB_NC_U16 : VOP3Inst_t16 <"v_sub_nc_u16", VOP_I16_I16_I16, sub>;
 
 } // End SubtargetPredicate = isGFX10Plus
 
+let True16Predicate = NotHasTrue16BitInsts, OtherPredicates = [isGFX10Plus] in {
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/VOP3Instructions.td:880]:
```suggestion
let True16Predicate = NotHasTrue16BitInsts, SubtargetPredicate = isGFX10Plus in {
```

File: llvm/lib/Target/AMDGPU/VOPInstructions.td [hunk 3]
```diff
@@ -111,7 +111,7 @@ class VOP3_Pseudo <string opName, VOPProfile P, list<dag> pattern = [],
 
   bit HasFP8DstByteSel = P.HasFP8DstByteSel;
 
-  let AsmOperands = !if(isVop3OpSel,
+  let AsmOperands = !if(!and(!not(P.IsTrue16), isVop3OpSel),
```
Comment (Sisyph) [inline llvm/lib/Target/AMDGPU/VOPInstructions.td:114]: I believe the intent is that we can remove AsmVOP3OpSel. When will we be able to do that? Can you make sure there is a comment about that on class getAsmVOP3OpSel ?

File: llvm/lib/Target/AMDGPU/VOPInstructions.td [hunk 4]
```diff
@@ -1342,17 +1500,43 @@ class Base_VOP3_DPP16<bits<10> op, VOP_DPP_Pseudo ps, string opName = ps.OpName>
   let OtherPredicates = ps.OtherPredicates;
 }
 
+class Base_VOP3_DPP16_t16<bits<10> op, VOP_DPP_Pseudo ps,
```
Comment (Sisyph) [inline llvm/lib/Target/AMDGPU/VOPInstructions.td:1503]: Can you replace the word Base in this // VOP3 DPP section to be another word? Maybe 'Intermediate'? This will allow us to follow the new convention at L784 where Base means the class is shared between T16 and regular.

File: llvm/lib/Target/AMDGPU/VOPInstructions.td [hunk 5]
```diff
@@ -1705,4 +1933,4 @@ def VOPTrue16Table : GenericTable {
 
   let PrimaryKey = ["Opcode"];
   let PrimaryKeyName = "getTrue16OpcodeHelper";
-}
```
Comment (Sisyph) [inline llvm/lib/Target/AMDGPU/VOPInstructions.td:1708]: Remove whitespace change

File: llvm/test/MC/AMDGPU/gfx12_asm_opsel.s [hunk 6]
```diff
@@ -0,0 +1,48 @@
+// RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -mattr=+wavefrontsize32,+real-true16 -show-encoding %s | FileCheck --check-prefixes=GFX12 %s
```
Comment (Sisyph) [inline llvm/test/MC/AMDGPU/gfx12_asm_opsel.s:1]: It looks like in gfx11, we have moved the tests with opsel from gfx11_asm_vop3.s to gfx11_asm_opsel.s. That seems ok to me. In gfx12 though, we still have the op_sel tests in gfx12_asm_vop3.s. So perhaps this test file is not needed?

Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx12_asm_opsel.s:1]: Should I move remove op_sel from gfx12_asm_vop3.s just like in gfx11?

File: llvm/lib/Target/AMDGPU/VOPInstructions.td (hunk 5 above)
Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/VOPInstructions.td:1708]: Hmmm I don't think I change anything here. Not sure why there is a diff

File: llvm/lib/Target/AMDGPU/VOPInstructions.td (hunk 4 above)
Comment (Sisyph) [reply llvm/lib/Target/AMDGPU/VOPInstructions.td:1503]: Maybe 'Helper' I like Helper better than Type or Class.

File: llvm/test/MC/AMDGPU/gfx12_asm_opsel.s (hunk 6 above)
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx12_asm_opsel.s:1]: Talked offline. I will remove this file first since it save me efforts for the upstreaming. I can move the gfx12 op_sel test to this seperate file later after all upstreaming has been done

File: llvm/lib/Target/AMDGPU/VOPInstructions.td (hunk 3 above)
Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/VOPInstructions.td:114]: added a comment in class getAsmVOP3OpSel

File: llvm/lib/Target/AMDGPU/VOPInstructions.td (hunk 4 above)
Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/VOPInstructions.td]: done

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp16.txt [hunk 7]
```diff
@@ -4115,88 +4115,276 @@
...
+0x05,0x10,0x03,0xd7,0xfa,0x04,0x02,0x00,0x01,0x60,0x01,0x13
+
+# W32-REAL16: v_add_nc_u16_e64_dpp v255.h, v255.l, v255.l op_sel:[0,0,1] clamp row_xmask:15 row_mask:0x3 bank_mask:0x0 bound_ctrl:1 fi:1 ; encoding: [0xff,0xc0,0x03,0xd7,0xfa,0xfe,0x03,0x00,0xff,0x6f,0x0d,0x30]
+# W32-FAKE16: v_add_nc_u16_e64_dpp v255, v255, v255 op_sel:[0,0,1] clamp row_xmask:15 row_mask:0x3 bank_mask:0x0 bound_ctrl:1 fi:1 ; encoding: [0xff,0xc0,0x03,0xd7,0xfa,0xfe,0x03,0x00,0xff,0x6f,0x0d,0x30]
+# W64-REAL16: v_add_nc_u16_e64_dpp v255.h, v255.l, v255.l op_sel:[0,0,1] clamp row_xmask:15 row_mask:0x3 bank_mask:0x0 bound_ctrl:1 fi:1 ; encoding: [0xff,0xc0,0x03,0xd7,0xfa,0xfe,0x03,0x00,0xff,0x6f,0x0d,0x30]
+# W64-FAKE16: v_add_nc_u16_e64_dpp v255, v255, v255 op_sel:[0,0,1] clamp row_xmask:15 row_mask:0x3 bank_mask:0x0 bound_ctrl:1 fi:1 ; encoding: [0xff,0xc0,0x03,0xd7,0xfa,0xfe,0x03,0x00,0xff,0x6f,0x0d,0x30]
+0xff,0xc0,0x03,0xd7,0xfa,0xfe,0x03,0x00,0xff,0x6f,0x0d,0x30
+
+# GFX11: v_add_nc_u16_e64_dpp v5.h, v1.h, v2.h op_sel:[1,1,1] row_share:0 row_mask:0xf bank_mask:0xf ; encoding: [0x05,0x58,0x03,0xd7,0xfa,0x04,0x02,0x00,0x01,0x50,0x01,0xff]
```
Comment (Sisyph) [inline llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp16.txt:4362]: Stray GFX11 CHECK. Please look at other files too.

**Review by Sisyph — COMMENTED** [review]
The commit message should also say it adds encoding info for VOP3 True16 including DPP and DPP8.

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp16.txt (hunk 7 above)
Comment (broxigarchen) [reply llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp16.txt]: done

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td [hunk 8]
```diff
@@ -760,9 +756,14 @@ class OpSelBinOpClampPat<SDPatternOperator node,
   (inst $src0_modifiers, $src0, $src1_modifiers, $src1, DSTCLAMP.ENABLE, 0)
 >;
 
-def : OpSelBinOpClampPat<saddsat, V_ADD_I16_e64>;
-def : OpSelBinOpClampPat<ssubsat, V_SUB_I16_e64>;
-} // End SubtargetPredicate = isGFX9Plus
+let SubtargetPredicate = isGFX9Plus, True16Predicate = NotHasTrue16BitInsts in {
+  def : OpSelBinOpClampPat<saddsat, V_ADD_I16_e64>;
+  def : OpSelBinOpClampPat<ssubsat, V_SUB_I16_e64>;
+} // End OtherPredicates = [isGFX9Plus], True16Predicate = NotHasTrue16BitInsts
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/VOP3Instructions.td:762]: Change hasn't reflected in the comment statement.

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td [hunk 9]
```diff
@@ -871,21 +872,31 @@ let SubtargetPredicate = isGFX10Plus in {
...
+let True16Predicate = NotHasTrue16BitInsts, SubtargetPredicate = isGFX10Plus in {
+   def : OpSelBinOpClampPat<uaddsat, V_ADD_NC_U16_e64>;
+   def : OpSelBinOpClampPat<usubsat, V_SUB_NC_U16_e64>;
+   // Undo sub x, c -> add x, -c canonicalization since c is more likely
+   // an inline immediate than -c.
+   def : GCNPat<
+     (add i16:$src0, (i16 NegSubInlineIntConst16:$src1)),
+     (V_SUB_NC_U16_e64 0, VSrc_b16:$src0, 0, NegSubInlineIntConst16:$src1, 0, 0)
+   >;
+} // End True16Predicate = NotHasTrue16BitInsts, OtherPredicates = [isGFX10Plus]
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/VOP3Instructions.td:889]: Ditto.

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td (hunk 8 above)
Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/VOP3Instructions.td]: done

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td (hunk 9 above)
Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/VOP3Instructions.td]: done

File: llvm/lib/Target/AMDGPU/VOPInstructions.td (hunk 5 above)
Comment (Sisyph) [reply llvm/lib/Target/AMDGPU/VOPInstructions.td:1708]: Perhaps clang-format? It seems to have deleted the blank line. I don't see a clear answer on whether a EoF blank line is required for tablegen, but would prefer to keep it.

**Review by Sisyph — APPROVED** [review]
LGTM, but please wait a day for others to have a look.

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td [hunk 10]
```diff
@@ -760,9 +756,14 @@ class OpSelBinOpClampPat<SDPatternOperator node,
...
-def : OpSelBinOpClampPat<ssubsat, V_SUB_I16_e64>;
-} // End SubtargetPredicate = isGFX9Plus
+let SubtargetPredicate = isGFX9Plus, True16Predicate = NotHasTrue16BitInsts in {
+  def : OpSelBinOpClampPat<saddsat, V_ADD_I16_e64>;
+  def : OpSelBinOpClampPat<ssubsat, V_SUB_I16_e64>;
+} // End SubtargetPredicate = isGFX9Plus, True16Predicate = NotHasTrue16BitInsts
+let True16Predicate = UseFakeTrue16Insts in {
+  def : OpSelBinOpClampPat<saddsat, V_ADD_I16_fake16_e64>;
+  def : OpSelBinOpClampPat<ssubsat, V_SUB_I16_fake16_e64>;
+} // End True16Predicate = UseFakeTrue16Insts
```
Comment (kosarev) [inline llvm/lib/Target/AMDGPU/VOP3Instructions.td:766]: Why don't these fake16 patterns need a subtarget predicate?

File: llvm/lib/Target/AMDGPU/SIInstrInfo.td [hunk 11]
```diff
@@ -2149,6 +2149,8 @@ class getAsmVOP3P <int NumSrcArgs, bit HasModifiers,
   string ret = dst#", "#src0#src1#src2#opsel#mods#clamp;
 }
 
+//FIXME-TRUE16 AsmVOP3OpSel will be deprecated after all
```
Comment (kosarev) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.td:2152]: Nit: space after //.

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td [hunk 12]
```diff
@@ -1104,6 +1115,17 @@ multiclass VOP3_Realtriple_with_name_gfx11_gfx12<bits<10> op, string opName,
 multiclass VOP3Dot_Realtriple_gfx11_gfx12<bits<10> op> :
   VOP3Dot_Realtriple<GFX11Gen, op>, VOP3Dot_Realtriple<GFX12Gen, op>;
 
+multiclass VOP3_Realtriple_t16_gfx11_gfx12<bits<10> op, string asmName, string opName = NAME,
+                                     string pseudo_mnemonic = "", bit isSingle = 0> :
```
Comment (kosarev) [inline llvm/lib/Target/AMDGPU/VOP3Instructions.td:1119]: Nit: indentation is off. +Same below.

File: llvm/lib/Target/AMDGPU/VOPInstructions.td [hunk 13]
```diff
@@ -1273,6 +1383,32 @@ class VOP3_Profile<VOPProfile P, VOP3Features Features = VOP3_REGULAR> : VOP3_Pr
 
 }
 
+class VOP3_Profile_True16<VOPProfile P, VOP3Features Features = VOP3_REGULAR> : VOPProfile_True16<P> {
+
```
Comment (kosarev) [inline llvm/lib/Target/AMDGPU/VOPInstructions.td:1387]: Nit: extra empty line. +Same below.

File: llvm/lib/Target/AMDGPU/VOPInstructions.td [hunk 14]
```diff
@@ -1342,17 +1500,43 @@ class Base_VOP3_DPP16<bits<10> op, VOP_DPP_Pseudo ps, string opName = ps.OpName>
...
+class VOP3_DPP16_t16<bits<10> op, VOP_DPP_Pseudo ps, int subtarget,
+                 string opName = ps.OpName>
+    : VOP3_DPP16_t16_Helper<op, ps, opName>, SIMCInstr<ps.PseudoInstr, subtarget>;
 
 class VOP3_DPP16_Gen<bits<10> op, VOP_DPP_Pseudo ps, GFXGen Gen,
-                     string opName = ps.OpName> :
-  VOP3_DPP16 <op, ps, Gen.Subtarget, opName> {
+                     string opName = ps.OpName>
+    : VOP3_DPP16 <op, ps, Gen.Subtarget, opName> {
```
Comment (kosarev) [inline llvm/lib/Target/AMDGPU/VOPInstructions.td:1527]: Whitespace change? Doesn't seem to match what we have downstream either.

File: llvm/lib/Target/AMDGPU/VOPInstructions.td [hunk 15]
```diff
@@ -1579,6 +1795,16 @@ multiclass VOP3beOnly_Realtriple<GFXGen Gen, bits<10> op> :
...
 
+// VOP1 and VOP2 depend on these triple defs
+
+multiclass VOP3_Realtriple_t16_gfx11<bits<10> op, string asmName, string opName = NAME,
+                                     string pseudo_mnemonic = "", bit isSingle = 0> :
+  VOP3_Realtriple_with_name<GFX11Gen, op, opName, asmName, pseudo_mnemonic, isSingle>;
+
+multiclass VOP3Only_Realtriple_t16_gfx11<bits<10> op, string asmName,
+                                     string opName = NAME, string pseudo_mnemonic = "">
+  : VOP3_Realtriple_t16_gfx11<op, asmName, opName, pseudo_mnemonic, 1>;
```
Comment (kosarev) [inline llvm/lib/Target/AMDGPU/VOPInstructions.td:1806]: Why was this moved here?

File: llvm/lib/Target/AMDGPU/VOPInstructions.td [hunk 16]
```diff
@@ -1610,6 +1832,16 @@ multiclass VOP3Only_Real_Base_gfx12<bits<10> op> :
 multiclass VOP3Only_Realtriple_t16_gfx12<bits<10> op> :
   VOP3Only_Realtriple<GFX12Gen, op>;
 
+multiclass VOP3_Realtriple_t16_gfx12<bits<10> op, string asmName, string opName = NAME,
+                                     string pseudo_mnemonic = "", bit isSingle = 0> :
+  VOP3_Realtriple_with_name<GFX12Gen, op, opName, asmName, pseudo_mnemonic, isSingle>;
+
+multiclass VOP3_Realtriple_t16_and_f16_gfx12<bits<10> op, string asmName, string opName = NAME,
```
Comment (kosarev) [inline llvm/lib/Target/AMDGPU/VOPInstructions.td:1839]: t16_and_fake16

**Review by kosarev — COMMENTED** [review]
This feel a lot for a single patch. Could this be a series of smaller changes?

File: llvm/lib/Target/AMDGPU/VOP3Instructions.td (hunk 10 above)
Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/VOP3Instructions.td:766]: UseFakeTrue16Insts indicates a post GFX11 target

Comment (kosarev) [reply llvm/lib/Target/AMDGPU/VOP3Instructions.td:766]: Right, it implies HasTrue16BitInsts. Thanks.

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3.s [hunk 17]
```diff
@@ -244,50 +244,59 @@ v_add_lshl_u32 v5, src_scc, vcc_lo, -1
...
+// GFX11: [0x05,0x00,0x0d,0xd7,0x6b,0xfe,0x01,0x00,0x0b,0xfe,0x00,0x00]
 
-v_add_nc_i16 v5, ttmp15, src_scc
-// GFX11: encoding: [0x05,0x00,0x0d,0xd7,0x7b,0xfa,0x01,0x00]
+v_add_nc_i16 v5.l, ttmp15, src_scc
+// GFX11: [0x05,0x00,0x0d,0xd7,0x7b,0xfa,0x01,0x00]
 
-v_add_nc_i16 v5, m0, 0.5
+v_add_nc_i16 v5.l, m0, 0.5
 // GFX11: encoding: [0x05,0x00,0x0d,0xd7,0x7d,0xe0,0x01,0x00]
```
[the remark this answers is not in the document]
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3.s:269]: This is coming from our downstreaming branch since I was running some script to diff and update the test file. 

I can add the encoding back but it probably will be a conflict when downstreaming

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3.s [hunk 18]
```diff
@@ -244,50 +244,59 @@ v_add_lshl_u32 v5, src_scc, vcc_lo, -1
...
-// GFX11: encoding: [0x05,0x00,0x0d,0xd7,0xc1,0xfe,0x00,0x00]
+v_add_nc_i16 v5.l, -1, exec_hi
+// GFX11: [0x05,0x00,0x0d,0xd7,0xc1,0xfe,0x00,0x00]
 
-v_add_nc_i16 v5, 0.5, m0 op_sel:[1,0,0]
+v_add_nc_i16 v5.l, 0.5, m0 op_sel:[1,0,0]
 // GFX11: encoding: [0x05,0x08,0x0d,0xd7,0xf0,0xfa,0x00,0x00]
 
-v_add_nc_i16 v5, src_scc, vcc_lo op_sel:[0,1,0]
-// GFX11: encoding: [0x05,0x10,0x0d,0xd7,0xfd,0xd4,0x00,0x00]
```
[the remark this answers is not in the document]
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3.s:287]: This went to the new file gfx11_asm_opsel.s

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3.s [hunk 19]
```diff
@@ -244,50 +244,59 @@ v_add_lshl_u32 v5, src_scc, vcc_lo, -1
...
+// GFX11: [0xff,0xc0,0x0d,0xd7,0xff,0xd6,0x00,0x00,0x0b,0xfe,0x00,0x00]
+
+v_add_nc_i16 v5.l, v1.h, v2.l
+// GFX11: [0x05,0x08,0x0d,0xd7,0x01,0x05,0x02,0x00]
+
+v_add_nc_i16 v5.l, v255.l, v255.h
+// GFX11: [0x05,0x10,0x0d,0xd7,0xff,0xff,0x03,0x00]
+
+v_add_nc_i16 v255.h, 0xfe0b, vcc_hi clamp
+// GFX11: [0xff,0xc0,0x0d,0xd7,0xff,0xd6,0x00,0x00,0x0b,0xfe,0x00,0x00]
```
[the remark this answers is not in the document]
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3.s:299]: This seems weird to me either. I think this might be a mistake I made when I am running the update script twice when copying the hi tests

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3.s (hunk 17 above)
[the remark this answers is not in the document]
Comment (kosarev) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3.s:269]: Right, so we either could add them downstream or remove them here with a separate change.

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3.s (hunk 18 above)
[the remark this answers is not in the document]
Comment (kosarev) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3.s:287]: Please make it a separate patch.

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s [hunk 20]
```diff
@@ -194,47 +194,59 @@ v_add_lshl_u32_e64_dpp v5, v1, v2, 0.5 row_xmask:0 row_mask:0x1 bank_mask:0x3 bo
...
+v_add_nc_i16_e64_dpp v5.l, v1.h, v2.l row_share:15 row_mask:0x0 bank_mask:0x1
+// GFX11: [0x05,0x08,0x0d,0xd7,0xfa,0x04,0x02,0x00,0x01,0x5f,0x01,0x01]
+
+v_add_nc_i16_e64_dpp v5.l, v1.l, v2.h row_xmask:0 row_mask:0x1 bank_mask:0x3 bound_ctrl:1 fi:0
+// GFX11: [0x05,0x10,0x0d,0xd7,0xfa,0x04,0x02,0x00,0x01,0x60,0x09,0x13]
+
+v_add_nc_i16_e64_dpp v255.h, v255.l, v255.l clamp row_xmask:15 row_mask:0x3 bank_mask:0x0 bound_ctrl:0 fi:1
+// GFX11: [0xff,0xc0,0x0d,0xd7,0xfa,0xfe,0x03,0x00,0xff,0x6f,0x05,0x30]
+
+v_add_nc_i16_e64_dpp v5.h, v1.h, v2.h row_share:0 row_mask:0xf bank_mask:0xf
```
[the remark this answers is not in the document]
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s:239]: No I think this is running the script twice. The whole block here are duplicated. Let me remove this

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s [hunk 21]
```diff
@@ -4475,30 +4523,6 @@ v_xor_b16_e64_dpp v5, v1, v2 row_xmask:0 row_mask:0x1 bank_mask:0x3 bound_ctrl:1
 v_xor_b16_e64_dpp v255, v255, v255 row_xmask:15 row_mask:0x3 bank_mask:0x0 bound_ctrl:0 fi:1
 // GFX11: [0xff,0x00,0x64,0xd7,0xfa,0xfe,0x03,0x00,0xff,0x6f,0x05,0x30]
 
-v_add_nc_i16_e64_dpp v5, v1, v2 op_sel:[1,1,1] row_share:0 row_mask:0xf bank_mask:0xf
-// GFX11: [0x05,0x58,0x0d,0xd7,0xfa,0x04,0x02,0x00,0x01,0x50,0x01,0xff]
```
[the remark this answers is not in the document]
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s:4479]: Ok. In this case I will put gfx11_asm_opsel.s back to the vop3.s. I'll also put up a patch in the downstream

File: llvm/lib/Target/AMDGPU/VOPInstructions.td (hunk 14 above)
Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/VOPInstructions.td:1527]: It should be matching the downstream I suppose

File: llvm/lib/Target/AMDGPU/VOPInstructions.td (hunk 15 above)
Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/VOPInstructions.td:1806]: I think it's to match downstream

File: llvm/test/MC/AMDGPU/gfx12_asm_vop3_dpp16.s [hunk 22]
```diff
@@ -214,47 +214,71 @@ v_add_lshl_u32_e64_dpp v5, v1, v2, 0.5 row_xmask:0 row_mask:0x1 bank_mask:0x3 bo
 v_add_lshl_u32_e64_dpp v255, v255, v255, src_scc row_xmask:15 row_mask:0x3 bank_mask:0x0 bound_ctrl:0 fi:1
 // GFX12: [0xff,0x00,0x47,0xd6,0xfa,0xfe,0xf7,0x03,0xff,0x6f,0x05,0x30]
 
-v_add_nc_i16_e64_dpp v5, v1, v2 quad_perm:[3,2,1,0]
-// GFX12: [0x05,0x00,0x0d,0xd7,0xfa,0x04,0x02,0x00,0x01,0x1b,0x00,0xff]
```
Comment (kosarev) [inline llvm/test/MC/AMDGPU/gfx12_asm_vop3_dpp16.s:218]: We just add some new cases here in front of the 0x05,0x00,0x0d,0xd7,0xfa,0x04,0x02,0x00,0x01,0x1b,0x00,0xff one, right?

File: llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3.txt [hunk 23]
```diff
@@ -189,49 +189,112 @@
...
-0x05,0x00,0x0d,0xd7,0x01,0x05,0x02,0x00
-
-# GFX11: v_add_nc_i16 v5, v255, v255             ; encoding: [0x05,0x00,0x0d,0xd7,0xff,0xff,0x03,0x00]
-0x05,0x00,0x0d,0xd7,0xff,0xff,0x03,0x00
-
-# GFX11: v_add_nc_i16 v5, s1, s2                 ; encoding: [0x05,0x00,0x0d,0xd7,0x01,0x04,0x00,0x00]
+# W32-REAL16: v_add_nc_i16 v5.l, v1.h, v2.l op_sel:[1,0,0] ; encoding: [0x05,0x08,0x0d,0xd7,0x01,0x05,0x02,0x00]
+# W32-FAKE16: v_add_nc_i16 v5, v1, v2 op_sel:[1,0,0]  ; encoding: [0x05,0x08,0x0d,0xd7,0x01,0x05,0x02,0x00]
+# W64-REAL16: v_add_nc_i16 v5.l, v1.h, v2.l op_sel:[1,0,0] ; encoding: [0x05,0x08,0x0d,0xd7,0x01,0x05,0x02,0x00]
+# W64-FAKE16: v_add_nc_i16 v5, v1, v2 op_sel:[1,0,0]  ; encoding: [0x05,0x08,0x0d,0xd7,0x01,0x05,0x02,0x00]
```
Comment (kosarev) [inline llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3.txt:195]: Should be GFX11-REAL16 and GFX11-FAKE16?

**Review by kosarev — APPROVED** [review]
LGTM with a few minor things to refine. If you expect this to not land smoothly downstream, especially the test changes, please make sure to prepare a downstream version of the patch before submitting here.

File: llvm/test/MC/AMDGPU/gfx12_asm_vop3_dpp16.s (hunk 22 above)
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx12_asm_vop3_dpp16.s:218]: yes

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s (hunk 21 above)
[the remark this answers is not in the document]
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s:4479]: Yes they are

File: llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3.txt (hunk 23 above)
Comment (broxigarchen) [reply llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3.txt:195]: I will prefer to use the old tag for now
